src/trait_/common_ord.rs

```diff
@@ -214,79 +231,131 @@ pub fn build_ord_signature(
 
 /// Builds order comparison recursively for all variants.
 #[cfg(not(feature = "nightly"))]
-fn build_recursive_order(
-	trait_: &DeriveTrait,
+fn build_discriminant_order(
+	repr: Option<Representation>,
+	item: &Item,
+	generics: &SplitGenerics<'_>,
 	variants: &[Data<'_>],
-	incomparable: &TokenStream,
+	path: &Path,
+	method: &TokenStream,
 ) -> TokenStream {
-	let mut less = quote! { ::core::cmp::Ordering::Less };
-	let mut greater = quote! { ::core::cmp::Ordering::Greater };
+	use std::{borrow::Cow, ops::Deref};
+
+	use proc_macro2::Span;
+	use syn::{parse_quote, Expr, ExprLit, Lit, LitInt};
+
+	let mut discriminants = Vec::<Cow<Expr>>::with_capacity(variants.len());
+	let mut has_non_isize = false;
+
+	for variant in variants {
+		let discriminant = if let Some(discriminant) = variant.discriminant {
+			if !has_non_isize
+				&& !matches!(
+					discriminant,
+					Expr::Lit(ExprLit {
+						lit: Lit::Int(_),
+						..
+					})
+				) {
```

Comment:
that does `-10` parse to? Could it be this: https://docs.rs/syn/latest/syn/struct.ExprUnary.html

Reply:
We could also, if we fancy, evaluate “binop-trees” on integers, as those would probably be the most like used, e.g., `1 << 4` or something like that.

Though, this is a stark complication, as we are basically implementing our own rust expression evaluator at that point.

Reply:
> that does `-10` parse to? Could it be this: https://docs.rs/syn/latest/syn/struct.ExprUnary.html

I don't know, we will find out after I add the tests.

> We could also, if we fancy, evaluate “binop-trees” on integers, as those would probably be the most like used, e.g., `1 << 4` or something like that.
> 
> Though, this is a stark complication, as we are basically implementing our own rust expression evaluator at that point.

I'm not sure we gain anything because the compiler is supposed to optimize `(1 << 4) + 1` away anyway.

Reply:
Yeah, this is only relevant for cases where this would allow us to use an `isize` instead of needing to fall back to an `i128`.

Reply:
Confirmed `-10` is parsed as `ExprUnary`.
Not sure if it makes a difference, will add more tests until I see what happens.

Reply:
I removed all expression parsing now.
The whole logic is much smaller and simplified now, we always fallback to `isize` if we are using the default representation. So this should be solid now.

Reply:
> Confirmed `-10` is parsed as `ExprUnary`. Not sure if it makes a difference, will add more tests until I see what happens.

Well if we only parse numbers `> i16::MIN` we wouldn't need to fall back to `i128`.

Reply:
I removed that as well, we don't do any parsing at all now.
See https://github.com/ModProg/derive-where/pull/77#issuecomment-1703012002 how this isn't a problem anymore.